Inject IJW Copy Constructor calls in the JIT instead of in the IL stubs

jkoritzinsky opened the pull request:
Fixes Issue #100751 and #100033 by injecting copy constructor calls in the JIT instead of injecting them in the IL and then using special hooks/interceptors to handle non-IL-supported/accessible scenarios.

Re-enable GS checks in Reverse P/Invokes now that we handel them correctly.

jkoritzinsky committed:
Revert "Disable GS checks in Reverse P/Invoke stubs to avoid missing these copies."

This reverts commit 79b1cd234fbb866cd34f6bc8ccad20578bf7ddc7.

jkoritzinsky committed:
Revert "Since the fix is in the CoreCLR VM side, we need to disable the tests in JitStress (as they won't be fixed when the JIT introduces these extra checks when not asking the VM)"

This reverts commit 35d0de9858ae7e78cbcd9fdecd3f41257b6f2d72.

jkoritzinsky committed:
Propagate the modreq to the stub and into the GS cookie logic to avoid creating shadow vars

jkoritzinsky committed:
Add RuntimeHelpers copy-constructor helper

jkoritzinsky committed:
Implement Reverse P/Invoke GS shadow copy for copy constructors

jkoritzinsky committed:
Use a jitinterface function instead of a jit helper so we don't need to do the generic method instantiation on every call (only when jitting)

jkoritzinsky committed:
Remove extra P/Invoke copy constructor call

jkoritzinsky committed:
Propagate the "special-copy" modifier into the JIT for byref locals (which the P/Invoke scenario will use)

jkoritzinsky committed:
Move copy constructor handling into the JIT for arguments

jkoritzinsky committed:
Remove now-dead code

jkoritzinsky committed:
Fix formatting

jkoritzinsky committed:
Fix build on Windows x64

AndyAyersMS commented:
I'm somewhat skeptical that annotating lcl vars in the jit will be sufficient. Isn't the need to invoke a copy ctor a property of the type and not of the variable?

jkoritzinsky committed:
Call copy helper in lowering and heavily reduce changes in higher layers

jkoritzinsky commented:
In our case, it actually is a property of the variable. In this particular scenario, the modifier means "this variable needs a copy constructor to be used whenever the value is copied, but the C++ compiler is not generating the body of this method. The runtime needs to emit copy constructor calls in this method."

jakobbotsch commented:
This is a bit too viral for my taste. It is basically adding a new type of local in the JIT that cannot be bitwise copied. All current and future transformations within the JIT need to account for this kind of local with its special rules. Also, it does not seem right to me that internal copies for this type of local results in insertion of user visible calls. For example, simple primitive functions like `gtSplitTree` will result in additional calls to the copy constructor.

I think instead we should restrict the insertion of the copy constructors to the few places we expect the copy to be user visible. Particularly we know that if the address of an IL local is taken, then it is going to be stable throughout the method. Because of that I do not think that internal copies made by the JIT need to call the copy constructor. Rather, we need to call the copy constructor at just the boundaries, before the address is stable and at the copy that the IL stub is inserting in the IL:
1. As you noticed, the shadow parameters results in user visible differences for the caller argument address and the callee argument address. I think the shadow parameters pass should be responsible for inserting the call to the copy constructor for this case. This is the boundary going into the function, where the local's address may change before being "published" to the IL. I don't think we have other places like that for address exposed variables.
2. For the copy in the IL, I assume it always comes in the form of the struct being pushed as an argument to the call. Can we instead track somewhere (either as a side table in the JIT or as information on the argument that we query the VM for) that the argument has an associated source IL local that the copy constructor must be invoked for? Then lowering would just call the copy constructor with `<address of argument>, LCL_ADDR <that local that we got from the VM/side table>`.

Of course this works based on the fact that the IL patterns here are very restricted, but I really would like to avoid that we try to teach the current JIT and all future changes to the JIT about the fact that types with copy constructors exist.

jkoritzinsky commented:
If I revert 390185cb244d3fb661a1f4b6197a7a852924ff45 then that moves the handling for shadow parameters and spills into the GS check/importer level. However, that still leaves the `PUTARG_STK` handling to happen in lowering.

If you have a recommendation for a better way to handle that case (and get it out of lowering) than how I'm currently doing it, I'm open for suggestions.

> Of course this works based on the fact that the IL patterns here are very restricted, but I really would like to avoid that we try to teach the current JIT and all future changes to the JIT about the fact that types with copy constructors exist.

I'm fine with only supporting restricted IL patterns, but I'd like suggestions of where to put asserts to ensure that any other IL patterns cause JIT asserts.

> Also, what about the destructor? When we insert one of these copy constructors it seems we inevitably also need to insert a destructor somewhere.

In this case (at least the way that it's traditionally been structured), when the IL stub does a copy, it actually does "copy constructor  of new + destructor of old", which is what I implemented `RuntimeHelpers.CopyConstruct<T>` to do in this case.

This provides us with the expected behavior as per our implicit contract with the C++/CLI compiler's codegen.

jakobbotsch commented:
> If you have a recommendation for a better way to handle that case (and get it out of lowering) than how I'm currently doing it, I'm open for suggestions.

I don't think we will be able to get it out of lowering. But what I'm after is that we make use of the fact that the only copies we allow inserting the copy constructor call for are the shadow params one and the stack pushed x86 one to the specific pinvoke call that we expect to see in this IL.
Other internal copies made by the JIT (whether it be CSE's, copies due to splitting nearby trees, promotion of some fields, introduction of temporaries for any reason) should not require any copy constructor invocations.

> I'm fine with only supporting restricted IL patterns, but I'd like suggestions of where to put asserts to ensure that any other IL patterns cause JIT asserts.

I'm specifically after allowing the JIT to make (bitwise) copies of these locals as long as it does not alter user visible behavior. Fundamentally once the address is taken of the IL local the JIT has to guarantee that it remains stable, and the transformations it makes cannot alter that visible behavior. 

Here is an example: say that your IL stub for some reason is passing a static field as the value of an argument that comes after the IL local with the copy constructor. The JIT is going to change your `M(vectorWithCopyConstructor, class.SomeStaticField)` into the equivalent of
```cpp
var copy = vectorWithCopyConstructor;
if (!class.IsInitialized)
  class.Cctor();
M(copy, class.SomeStaticField)
```

The same form of change can happen for a multitude of reasons internally within the JIT. I would expect that you can hit this particular introduction of a copy today when the JIT runs with `STRESS_SPLIT_TREES_RANDOMLY`.
With this PR, we will be invoking the copy constructor one more time when the JIT makes this internal transformation. However, I think we should instead have lowering recognize the call to `M` and introduce the copy constructor call with `<address of argument>, LCL_ADDR(vectorWithCopyConstructor)`, despite the fact that previous transformations within the JIT now means that the bits are in `copy` instead.

I don't think it is simple to assert anything here, but I don't think the introducing the copy constructor calls more liberally improves the situation. You can still create IL that does not properly maintain the "move" semantics you are implementing here (e.g. two copies from the same IL local would cause problems that would not result in any asserts here).

jkoritzinsky commented:
I think I understand better now. I'll try to adjust the implementation implement it with a table in the JIT to do the copy/move from the "original" local into the arg slot for the specific GT_PUTARG_STK case and do the copy/move into the shadow arg for the Reverse P/Invoke case.

jkoritzinsky committed:
Revert "Call copy helper in lowering and heavily reduce changes in higher layers"

This reverts commit 390185cb244d3fb661a1f4b6197a7a852924ff45.

jkoritzinsky committed:
Implement copy-constructor handling for P/Invokes via a map back to the original arg and don't do intermediate copy-constructor calls.

The bitwise copies are invalid, but there's no possible way for the user to see them. They'll always get passed a correctly copy-constructed value.

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -16604,6 +16604,18 @@ GenTree* Compiler::gtNewTempStore(
         val->gtType = valTyp;
     }
+
+    if (val->OperGet() == GT_BLK && val->AsBlk()->Addr()->OperGet() == GT_LCL_VAR)
+    {
+        GenTreeLclVarCommon* lclVar = val->AsBlk()->Addr()->AsLclVarCommon();
+        LclVarDsc* srcVarDsc = &lvaTable[lclVar->GetLclNum()];
+        if (srcVarDsc->lvRequiresSpecialCopy)
+        {
+            JITDUMP("Recording V%02u as an alias of the special-copy local V%02u\n", tmp, lclVar->GetLclNum());
+            GetSpecialCopyLocalsMap()->Emplace(tmp, lclVar->GetLclNum());
+        }
+    }
```
I think this lookup should be done from call + arg back to local. Otherwise we are adding a requirement that all copies in the JIT need to be created in a specific way and that this table is maintained for them all. I don't think we need to do that.

jakobbotsch commented on src/coreclr/jit/gschecks.cpp:
```diff
@@ -516,13 +516,48 @@ void Compiler::gsParamsToShadows()
-        GenTree* src = gtNewLclvNode(lclNum, varDsc->TypeGet());
-        src->gtFlags |= GTF_DONT_CSE;
-        GenTree* store = gtNewStoreLclVarNode(shadowVarNum, src);
+        if (varDsc->lvRequiresSpecialCopy)
```
```suggestion
        if (varDsc->lvRequiresSpecialCopy && (varDsc->TypeGet() == TYP_STRUCT))
```

jakobbotsch commented on src/coreclr/jit/gschecks.cpp:
```diff
@@ -516,13 +516,48 @@ void Compiler::gsParamsToShadows()
+            call->gtArgs.PushBack(this, NewCallArg::Primitive(src));
+
+            fgEnsureFirstBBisScratch();
+            compCurBB = fgFirstBB; // Needed by some morphing
+            if (opts.IsReversePInvoke())
+            {
+                JITDUMP(
+                    "Inserting special copy helper call at the end of the first block after Reverse P/Invoke transition\n");
+                // If we are in a reverse P/Invoke, then we need to insert
+                // the call at the end of the first block as we need to do the GC transition
+                // before we can call the helper.
+                (void)fgNewStmtAtEnd(fgFirstBB, fgMorphTree(call));
+            }
```
Hmm, I don't think there is any guarantee that there are no uses of the parameter before this point.
Can you add an assert? You can use `gtHasRef` on the root node of all the statements before this. Long term it seems like we ought to add the GC transitions later in the JIT.

jakobbotsch commented on src/coreclr/inc/icorjitinfoimpl_generated.h:
```diff
@@ -741,6 +741,9 @@ uint32_t getJitFlags(
           CORJIT_FLAGS* flags,
           uint32_t sizeInBytes) override;
 
+CORINFO_METHOD_HANDLE GetSpecialCopyHelper(
```
Most JIT-EE functions are named in camelCase.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -1879,6 +1950,12 @@ void Lowering::LowerArg(GenTreeCall* call, CallArg* callArg, bool late)
             ReplaceArgWithPutArgOrBitcast(ppArg, putArg);
         }
+
+        if (putArg->OperIsPutArgStk() && arg->OperIs(GT_LCL_VAR) &&
+            comp->lvaRequiresSpecialCopy(arg->AsLclVar()->GetLclNum()))
```
This is the point where we should instead map back from the call + arg to the source local. Otherwise we are adding long term requirement that we end up with a `LCL_VAR` node here after all JIT transformations.

jakobbotsch replied:
I would probably move the handling to a separate function that is called right before `LegalizeArgPlacement` in `LowerArgsForCall`.

Would it be too crude to do it something like this:
```cpp
if (IsILStub() && call->IsPInvoke())
{
  unsigned argIndex = 0;
  for (CallArg& arg : call->gtArgs.Args())
  {
    if (!arg.IsUserArg())
    {
      continue;
    }

    if (argIndex >= comp->info.compArgsCount)
    {
      break;
    }

    // check if parameter at the exact same index as the IL argument is marked as requiring special copy, assuming that it is being passed 1:1 to the pinvoke
    if (lvaRequiresSpecialCopy(argIndex))
    {
      InsertCopyConstructor(arg); 
    }
  }
}
```
?

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -1523,6 +1523,77 @@ void Lowering::ReplaceArgWithPutArgOrBitcast(GenTree** argSlot, GenTree* putArgO
+
+    GenTree* dest;
+
+#if FEATURE_FIXED_OUT_ARGS
+    dest = comp->gtNewOperNode(GT_ADD, TYP_I_IMPL, comp->gtNewPhysRegNode(REG_SPBASE, TYP_I_IMPL),
+                               comp->gtNewIconNode(putArgStk->getArgOffset()));
+#else
+    dest = comp->gtNewPhysRegNode(REG_SPBASE, TYP_I_IMPL);
+#endif
```
I think the `FEATURE_FIXED_OUT_ARGS` case is going to require more work than this. But also, for `FEATURE_FIXED_OUT_ARGS` we have an invariant that there are no calls between a `PUTARG_STK` and its corresponding `CALL` node (see `LegalizeArgPlacement`). How hard would it be to keep the non-x86 cases outside the JIT?

jakobbotsch replied:
Actually, we should be able to keep this invariant. I think the code just isn't working correctly for win-x64 as it is, since the destination here is not going to be the outgoing arg area (which this code computes), but rather the target of the implicit byref. So the code should instead be calling the copy constructor on same value as is being passed as the argument in the `FEATURE_FIXED_OUT_ARGS` case. I would add assert that the arg is `PassedByRef` as well.

jakobbotsch replied:
(Also, I would expect that a `PUTARG_REG` is possible in that case too.)

jkoritzinsky commented on src/coreclr/jit/lclvars.cpp:
```diff
@@ -326,6 +326,19 @@ void Compiler::lvaInitTypeRef()
             }
         }
 
+        if ((corInfoTypeWithMod & CORINFO_TYPE_MOD_COPY_WITH_HELPER) != 0)
+        {
+            if (corInfoType == CORINFO_TYPE_VALUECLASS || corInfoType == CORINFO_TYPE_BYREF)
+            {
+                JITDUMP("Setting lvRequiresSpecialCopy for V%02u\n", varNum);
+                varDsc->lvRequiresSpecialCopy = 1;
+            }
+            else
+            {
+                JITDUMP("Ignoring special copy for non-struct, non-byref V%02u\n", varNum);
+            }
+        }
+
```
Currently you can see it on a local because the IL stub uses a pinned byref local. I'll see if I can change the IL stub strategy to just use the local for pinning and use the arg for passing.

jkoritzinsky committed:
Various PR feedback and propagate special copy directly from the IL arguments to the arg slot instead of tracing through locals.

jkoritzinsky committed:
Use a simple bool array instead of a vector as a function that has one parameter of this type will likely have multiple.

jkoritzinsky committed:
Format

jkoritzinsky committed:
Update hashing

jkoritzinsky committed:
Remove assert

jkoritzinsky committed:
Update Misc test

jkoritzinsky committed:
Cast through size_t to make it explicit that we're dropping the top bits

jkoritzinsky committed:
Use compArgsCount and remove assert

jkoritzinsky committed:
Various fixes

jkoritzinsky committed:
Fix GCC build

jkoritzinsky committed:
Fix signature type check

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into copy-ctor-jit-inject

jakobbotsch commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.h:
```diff
@@ -898,6 +898,10 @@ class MethodContext
     void dmpGetStringConfigValue(DWORD nameIndex, DWORD result);
     const WCHAR* repGetStringConfigValue(const WCHAR* name);
 
+    void recgetSpecialCopyHelper(CORINFO_CLASS_HANDLE type, CORINFO_METHOD_HANDLE helper);
+    void dmpgetSpecialCopyHelper(DWORDLONG key, DWORDLONG value);
+    CORINFO_METHOD_HANDLE repgetSpecialCopyHelper(CORINFO_CLASS_HANDLE type);
```
Nit -- naming is not consistent with other functions

jakobbotsch commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.h:
```diff
@@ -1189,6 +1193,7 @@ enum mcPackets
     Packet_GetTypeForBoxOnStack = 221,
     Packet_GetTypeDefinition = 222,
     Packet_GetFpStructLowering = 223,
+    Packet_getSpecialCopyHelper = 224,
```
Ditto

jakobbotsch commented on src/coreclr/jit/compiler.h:
```diff
@@ -5703,6 +5703,29 @@ class Compiler
     const CORINFO_SWIFT_LOWERING* GetSwiftLowering(CORINFO_CLASS_HANDLE clsHnd);
 #endif
 
+    bool* m_specialCopyArgs;
+    bool recordArgRequiresSpecialCopy(unsigned argNum)
+    {
+        if (argNum >= info.compArgsCount)
+        {
+            return false;
+        }
+
+        if (m_specialCopyArgs == nullptr)
+        {
+            m_specialCopyArgs = new (getAllocator()) bool[info.compArgsCount];
+            memset(m_specialCopyArgs, 0, info.compArgsCount * sizeof(bool));
+        }
+
+        m_specialCopyArgs[argNum] = true;
+        return true;
+    }
```
Is this to avoid having the bit in `LclVarDsc`?

jkoritzinsky replied:
Yeah that was the idea. No need to pay for the extra space if this is only used in a very small subset of scenarios.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -1994,9 +1994,107 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
+                continue;
+            }
+
+            if (argIndex >= comp->info.compArgsCount)
+            {
+                break;
+            }
+
+            // check if parameter at the same index as the IL argument is marked as requiring special copy, assuming
+            // that it is being passed 1:1 to the pinvoke
+            unsigned paramIndex = comp->compMap2ILvarNum(argIndex);
+            if ((paramIndex != (unsigned)ICorDebugInfo::UNKNOWN_ILNUM) && comp->argRequiresSpecialCopy(paramIndex) &&
+                arg.GetSignatureType() == TYP_STRUCT)
```
Sorry, I think I gave you the wrong helper function to use here. I think this should rather be:

```suggestion
            unsigned paramLclNum = comp->compMapILargNum(argIndex);
            assert(paramLclNum < comp->info.compArgsCount);

            // check if parameter at the same index as the IL argument is marked as requiring special copy, assuming
            // that it is being passed 1:1 to the pinvoke
            if (comp->argRequiresSpecialCopy(paramLclNum) && (arg.GetSignatureType() == TYP_STRUCT))
```

jakobbotsch replied:
How is win-x64 handled in the new scheme? It looks like the JIT is not handling it. (I expect the `GT_PUTARG_STK` assert would be hit below otherwise.)

jkoritzinsky replied:
win-x64 ends up passing pointers instead of structs. That's where the GetSignatureType came from (before that we hit the assert).

jakobbotsch replied:
Can you put the code to handle inserting the copy constructor calls under `TARGET_X86` then?

jkoritzinsky committed:
Feedback

JulieLeeMSFT commented:
@jkoritzinsky please check test failures.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -1994,9 +1994,100 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
+    var_types lclType = genActualType(comp->lvaGetDesc(lclNum));
+    if (lclType == TYP_BYREF || lclType == TYP_I_IMPL)
+    {
+        src = comp->gtNewLclVarNode(lclNum, TYP_I_IMPL);
```
```suggestion
        src = comp->gtNewLclVarNode(lclNum, lclType);
```

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -1994,9 +1994,100 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
+    GenTree* dest = comp->gtNewPhysRegNode(REG_SPBASE, TYP_I_IMPL);
+
+    GenTree*  src;
+    var_types lclType = genActualType(comp->lvaGetDesc(lclNum));
```
```suggestion
    var_types lclType = comp->lvaGetDesc(lclNum);
```
(Not necessary if we expect this to always be a pointer/byref/struct)

jakobbotsch reviewed (approved):
JIT changes LGTM, with a few minor things.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -1994,9 +1994,100 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
+    }
+}
+
+void Lowering::InsertSpecialCopyArg(GenTreePutArgStk* putArgStk, CORINFO_CLASS_HANDLE argType, unsigned lclNum)
```
Would be nice to add function header here and above to give an overview of what is going on here and the special assumptions we make on the IL semantics for these stubs.

jkoritzinsky committed:
Add headers

jkoritzinsky committed:
Fix superpmi build failures. Correctly handle multiple-arg copy constructor scenarios (by counting args in the correct direction)

jkoritzinsky commented:
@jakobbotsch I had to make a few changes in `Lowering::LowerSpecialCallArgs` because of how arguments are ordered in unmanaged calls. If you could take one more look, I'd appreciate it.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -2002,21 +2002,56 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
                 continue;
             }
 
+            if (call->GetUnmanagedCallConv() == CorInfoCallConvExtension::Thiscall &&
+                argIndex == call->gtArgs.CountUserArgs() - 1)
+            {
+                assert(arg.GetNode()->OperIs(GT_PUTARG_REG));
+                continue;
+            }
+
+            if (argIndex >= comp->info.compILargsCount)
+            {
+                argIndex--;
+                continue;
+            }
```
I think you need some handling for retbufs as well.

jakobbotsch replied:
Ah, I guess not... Should be skipped by `IsUserArg`

jakobbotsch replied:
When is this `argIndex >= comp->info.compILargsCount` true?

jkoritzinsky replied:
This fails on the LCID interop tests, in which we introduce an additional argument into the native call that isn't present in the managed P/Invoke signature.

The C++/CLI compiler will never emit a P/Invoke that uses both copy constructors and introduces the LCID argument.

jakobbotsch replied:
I guess one way to be completely explicit about it would be to check `m_specialCopyArgs != nullptr` above and add an `assert(call.CountUserArgs() == comp->info.compILargsCount)`, but this looks ok to me as well.

jkoritzinsky replied:
I'll add the assert. Better to be explicit about it, and adding that check will probably help jit throughput (as we won't iterate through the args unless we actually are in a scenario where we may do something).

jkoritzinsky replied:
I'll add the assert. Better to be explicit about it, and adding that check will probably help jit throughput (as we won't iterate through the args unless we actually are in a scenario where we may do something).

jkoritzinsky committed:
Add check and assert

jkoritzinsky commented:
/azp run runtime, runtime-dev-innerloop
